properties/pkg/properties.go

```diff
@@ -0,0 +1,232 @@
+// Copyright © 2021 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package properties
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+)
+
+const (
+	// Default delimiter used for exporting Property
+	DefaultSeparator = "="
+)
+
+type keyIndex struct {
+	key   string
+	index uint
+}
+
+type keyIndexList []keyIndex
+
+func (k keyIndexList) Len() int {
+	return len(k)
+}
+
+func (k keyIndexList) Swap(i, j int) {
+	k[i], k[j] = k[j], k[i]
+}
+
+func (k keyIndexList) Less(i, j int) bool {
+	return k[i].index < k[j].index
+}
+
+// Properties is used to store a group of Property items belong together.
+// It also supports number of operations on the stored Property items.
+type Properties struct {
+	// Map of Property objects
+	properties map[string]Property
+	// Map of Property keys and indices for bookkeeping  their original
+	// order in the Properties document.
+	keys map[string]keyIndex
+	// Index for the next Property to be added to Properties
+	nextKeyIndex uint
+	// State lock to prevent concurrent updates of Properties
+	mutex sync.RWMutex
+}
+
+// Get a Property using by its name.
+func (p *Properties) Get(key string) (Property, bool) {
+	propByKey, found := p.properties[key]
+	if !found {
+		return Property{}, false
+	}
+	return propByKey, true
+}
+
+// Get list of Property names.
+func (p *Properties) Keys() []string {
+	// Acquire read lock
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+
+	// Create an keyIndexList with the size of the keys map
+	keyIdxList := make(keyIndexList, len(p.keys))
+
+	// Add keyIndex items from keys map to keyIndexList before sorting
+	idx := 0
+	for _, keyIdx := range p.keys {
+		keyIdxList[idx] = keyIdx
+		idx++
+	}
+
+	// Sort keys in keyIndexList by their index
+	sort.Sort(keyIdxList)
+
+	// Create a string slice with the size of the keys map
+	// holding the properties keys in order
+	keys := make([]string, len(p.keys))
+
+	// Retrieve the keys from the sorted keyIndexList
+	l := len(keyIdxList)
+	for idx := 0; idx < l; idx++ {
+		keys[idx] = keyIdxList[idx].key
+	}
+
+	return keys
```

Comment:
Using cap instead of len makes this block a bit more concise.
```suggestion
	// Create an keyIndexList with the size of the keys map
	keyIdxList := make(keyIndexList, 0, len(p.keys))

	// Add keyIndex items from keys map to keyIndexList before sorting
	for _, keyIdx := range p.keys {
		keyIdxList = append(keyIdxList, keyIdx)
	}

	// Sort keys in keyIndexList by their index
	sort.Sort(keyIdxList)

	// Create a string slice with the size of the keys map
	// holding the properties keys in order
	keys := make([]string, 0, len(p.keys))

	// Retrieve the keys from the sorted keyIndexList
	for _, keyIdx := range keyIdxList {
	        keys = append(keys, keyIdx.key)
	}

	return keys
```